Add support for CSS modules with SSR consistency

aduth opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of ARC-85
Related to / builds upon #103089

## Proposed Changes

* Extends default CSS modules support to ensure consistent class name generation between client-side and server-side builds
* Converts an example component as demonstration of server-rendered behavior (`MarketplaceFooter`)

Our default Webpack configuration already has support for CSS modules because [it is enabled by default in `css-loader`](https://www.npmjs.com/package/css-loader#modules), but it requires a few additions:

- `css-loader` [generates a unique salt](https://github.com/webpack-contrib/css-loader/blob/961bdfa2ce8c16ff4e0f2de43a69923610e3c0d8/src/utils.js#L341-L345) that can produce different results between builds. Therefore, we override `getLocalIdent` to produce a stable result
- We only want the server to generate an object mapping of original -> transformed class names and not fully embed the CSS, so we use [`exportOnlyLocals`](https://www.npmjs.com/package/css-loader#exportonlylocals) and disable `MiniCssExtractPlugin`
- Our current server Webpack configuration [explicitly ignores](https://github.com/Automattic/wp-calypso/blob/c44eb9c8686813bfdd525de5347622f076f8e9f8/client/webpack.config.node.js#L130-L133) Sass/CSS files, so we need to exempt CSS module files from that ignore pattern

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Support adoption of CSS modules within paths leveraging server rendering
* Benefits of component conversion:
   * Reference demonstration of using CSS modules
   * Includes some workarounds for common use-cases (conditional styles, styles overriding descendent components)
   * Migrate _away_ from Emotion styles
   * Clearer separation of `MarketplaceFooter` from `EducationFooter`: the previous location had two component definitions serving two distinct purposes

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Verify no visual difference in the marketplace footer

1. Go to a plugin details page
   - Local development: http://calypso.localhost:3000/plugins/sensei-pro
   - Production: https://wordpress.com/plugins/sensei-pro
2. Observe the footer at the bottom of the page is consistent between environments
3. Repeat steps while logged out (e.g. using a private browsing window). The visual appearance is slightly different between logged-in and logged-out users (vertical padding, background color)

Logged In|Logged Out
---|---
![image](https://github.com/user-attachments/assets/fc3939e9-b016-480b-a9dc-749270cd3b1d)|![image](https://github.com/user-attachments/assets/bf8e0059-41ae-475d-8d97-1b46c180a053)



Ensure that generated class names are consistent:

1. Run `yarn build-server`
2. Run `yarn build-client`
3. Search for generated server class name of one of the components:
   - `grep marketplace-footer__three-column build/server.js`
   - Example: `"marketplace-footer__three-column": `_02216b3f2e``
5. Search for generated client class name of the same component:
   - `grep -R marketplace-footer__three-column public/evergreen`
   - Example: `\"marketplace-footer__three-column\":\"_02216b3f2e\"`
6. Observe that the class names are identical


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Add support for CSS modules with SSR consistency

aduth committed:
Convert plugins MarketplaceFooter to use CSS modules

Demonstrate functional server-rendered CSS modules behavior

mirka commented on client/my-sites/plugins/marketplace-footer/index.tsx:
```diff
@@ -0,0 +1,65 @@
+			>
+				{ ( ! isLoggedIn || currentUserSiteCount === 0 ) && (
+					<Button
+						className={ `is-primary ${ styles[ 'marketplace-footer__cta' ] }` }
```
Preexisting issue, but looks like this should use the `primary` prop and not the `is-primary` class directly.

https://github.com/Automattic/wp-calypso/blob/a82a1292b0d2c0bff0ede8a6e555c324b35e1154/packages/components/src/button/index.tsx#L78

aduth replied:
> Preexisting issue, but looks like this should use the `primary` prop and not the `is-primary` class directly.
> 
> https://github.com/Automattic/wp-calypso/blob/a82a1292b0d2c0bff0ede8a6e555c324b35e1154/packages/components/src/button/index.tsx#L78

Fixed in 65cb83006f5

mirka commented on client/my-sites/plugins/education-footer/index.tsx:
```diff
@@ -107,51 +75,6 @@ const CardText = styled.span< { color: string } >`
-						{ __( 'Get Started' ) }
-					</Button>
-				) }
-				<ThreeColumnContainer>
```
I was almost about to say we should keep the `ThreeColumnContainer` in sync for the `EducationFooter` and `MarketplaceFooter`, but looking at the actual components they look pretty unrelated. Probably better that they're separated now.

mirka commented on client/my-sites/plugins/marketplace-footer/styles.module.css:
```diff
@@ -0,0 +1,37 @@
+    }
+}
+
+.marketplace-footer__cta {
+	min-width: 122px;
+	margin-bottom: 26px;
+}
+
+@media ( max-width: 660px ) {
+	.marketplace-footer__cta {
+		margin-left: 16px;
+		margin-right: 16px;
+	}
+}
```
Any reason we're not nesting the media queries in the selector? I find these harder to grok when skimming.

aduth replied:
> Any reason we're not nesting the media queries in the selector? I find these harder to grok when skimming.

To be honest, I wasn't sure if nesting media queries was supported with plain-CSS nesting (this is not `.scss`), but [confirmed that it is](https://drafts.csswg.org/css-nesting/#conditionals).

aduth replied:
> Any reason we're not nesting the media queries in the selector? I find these harder to grok when skimming.

Updated to use nesting in e4229426575.

mirka commented on client/my-sites/plugins/marketplace-footer/styles.module.css:
```diff
@@ -0,0 +1,37 @@
+	margin-bottom: -32px;
+
+    & .marketplace-footer__section::before {
+        background-color: #f6f7f7;
```
Probably meant to be this and not a magic number:

```suggestion
        background-color: var(--studio-gray-0);
```

aduth replied:
> Probably meant to be this and not a magic number:

Updated in 1e69f46d532.

mirka commented on client/my-sites/plugins/marketplace-footer/styles.module.css:
Nit: There are some aberrations, but I think the standard file naming in the repo is supposed to be `style` singular, not plural.

aduth replied:
> Nit: There are some aberrations, but I think the standard file naming in the repo is supposed to be `style` singular, not plural.

Updated in 06538607f4c.

mirka commented on client/my-sites/plugins/marketplace-footer/styles.module.css:
```diff
@@ -0,0 +1,37 @@
+	--color-accent-60: var( --studio-blue-60 );
+	margin-bottom: -32px;
+
+    & .marketplace-footer__section::before {
```
The `.marketplace-footer__section` selectors don't need to be nested anymore since we have our own unique class.

aduth replied:
> The `.marketplace-footer__section` selectors don't need to be nested anymore since we have our own unique class.

While that's true, we have an issue that these styles are meant to [override the default styles applied by the `Section` component](https://github.com/Automattic/wp-calypso/blob/47dd847a9587287df79e6e6036ab37d6f06d4965/client/components/section/index.tsx#L31-L32). If we un-nest this, the selectors will have equal specificity, which may mean that the override doesn't take effect.

A few thoughts on how we could address this:

- Normally I might want to do something like `.marketplace-footer__section.section::before` to both explicitly make it clear that we're overriding styles from `.section` and increase the specificity, but since `Section` uses Emotion, we don't have a stable selector to use.
- Ideally we wouldn't be overriding through styles like this in the first place, and `Section` would allow (or not allow) what types of customizations it expects to be permitted for the background color. In a way, it could be argued that it's "by design" that the background isn't allowed to be overridden through styles like this, and this is part of what we hope to achieve with using CSS modules.
- Absent the other options, we could leave this as-is.

aduth replied:
I think I'm inclined to leave this as-is for the purpose of this pull request, and follow-up with an enhancement to the `<Section />` component to better support additional variations of background color.

mirka commented on client/my-sites/plugins/marketplace-footer/styles.module.css:
```diff
@@ -0,0 +1,37 @@
+        background-color: #f6f7f7;
+    }
+
+    &:not(.is-logged-in) .marketplace-footer__section {
```
Looks like the modifier could've been set directly on `marketplace-footer__section` and not on `marketplace-footer`.

aduth replied:
> Looks like the modifier could've been set directly on `marketplace-footer__section` and not on `marketplace-footer`.

This sent me down a rabbit hole on trying to determine if modifier classes are valid on BEM **E**lements 😂 Since I usually only see them paired with **B**locks. But it seems to be perfectly valid!

aduth replied:
> Looks like the modifier could've been set directly on `marketplace-footer__section` and not on `marketplace-footer`.

Updated in fd65ebf19ef.

mirka reviewed (approved):
Looks great, and I've confirmed that the output is stable as expected.

Didn't mean to nitpick the CSS — I don't think we necessarily need to clean up every possible thing when converting existing code, but just left some thoughts in case it  adds to the pile of patterns that _can_ be simplified when converted to modules. In fact, for "quick" conversions, it may actually be safer and efficient to maintain the existing cascade/specificity as much as possible.

aduth committed:
Use primary prop instead of hard-coded class name

See: https://github.com/Automattic/wp-calypso/pull/103606#discussion_r2105441682

aduth committed:
Use nesting for media query styles

See: https://github.com/Automattic/wp-calypso/pull/103606#discussion_r2105467059

aduth committed:
Use color palette variable instead of hardcoded value

See: https://github.com/Automattic/wp-calypso/pull/103606#discussion_r2105472117

aduth committed:
Rename stylesheet as style.module.css

See: https://github.com/Automattic/wp-calypso/pull/103606#discussion_r2105475788

aduth committed:
Apply modifier to child element to reduce nesting

jsnajdr commented on packages/calypso-build/webpack/sass.js:
```diff
@@ -8,16 +9,27 @@ const MiniCSSWithRTLPlugin = require( './mini-css-with-rtl' );
+						createHash( 'md5' )
+							.update( context.resourcePath + localName )
+							.digest( 'hex' )
+							.substr( 0, 10 ),
```
This currently generates completely opaque class names that begin with `_`:
```
const style_module = ({
  "marketplace-footer": `_5df97c7b40`,
  "marketplace-footer__section": `_3c3498323c`,
  "is-logged-in": `_d93f9bf602`,
  "marketplace-footer__cta": `_92b887b139`,
  "marketplace-footer__three-column": `_bf9c1e28f8`
});
```
Could we make the styles (and the DOM markup that the React app generates) more human readable by adding the hash only as a suffix to the original name? The class names then would be like
```
marketplace-footer_5df97c7b40
```
Without this, inspecting DOM markup in devtools would be almost impossible: everything would be a `div` with unreadable class names.

aduth replied:
To some extent, the opaqueness is the point, since I believe that one of the goals of using CSS modules is to close components for external extension in order to avoid CSS (class names, etc.) being considered as part of a component's public API for backwards-compatibility.

Using the recommendation you suggested, it would open up the possibility of people doing something like...

```scss
.my-component [class^="marketplace-footer_"] {
  /* ...overrides */
}
```

But I agree it does have some trade-offs, and I've even caught myself being confused by the DevTools DOM inspector structure trying to find where styles are applied.

A couple thoughts:

- Maybe we don't need to be so strictly rigid here, and someone overriding like above is either unlikely or requires some tacit understanding of the overrides being fragile
- We could change the behavior to fully mangle in production builds and have more-helpful names in development. This could be error-prone however, since code like above would work in development and fail when deployed to production. Additionally, it could be helpful to maintain some debuggability in production.
- If mangling names is beneficial for backwards-compatibility, maybe we limit it to applying to situations where we need to more carefully consider backwards-compatibility (`packages/components`, or packages more broadly).

aduth replied:
Another idea is that we could have Stylelint check and lint against the user of `[class^=...]`. It's pretty uncommon and almost always a code smell to at least lint by default and require explicit override to disable.

I think I'll incorporate the component basename.

aduth replied:
I updated in 9ff1d89 to incorporate the original identifier, so now we get something like:

```ts
const style_module = ({
	"marketplace-footer": `marketplace-footer_de608`,
	"marketplace-footer__section": `marketplace-footer__section_f95a5`,
	"is-logged-in": `is-logged-in_91412`,
	"marketplace-footer__cta": `marketplace-footer__cta_12292`,
	"marketplace-footer__three-column": `marketplace-footer__three-column_d7cfb`
});
```

One other thing I noticed is that we want the string to be deterministic so that it's the same between client and server builds, but not _too_ deterministic that the class name will include a hash that is relatively stable over time (e.g. allowing someone to just target `.marketplace-footer_de608` for overrides). 

Open to suggestions, but I opted to try to detect the current git HEAD SHA value and use that as part of the generated hash. If it's unavailable for any reason (e.g. static file copy of the project), it'll gracefully fall back to creating a hash like `hash.update( 'marketplace-footerundefined' )`.

ciampo replied:
This sounds like a good compromise. We can always revisit at a later point (especially given that it's not meant to be a public API contract).

I also like the idea of a Stylelint check, although we need to make sure not to create too many false positives.

aduth committed:
Incorporate original localName in generated class name

Also use git head sha (if available) to have value be deterministic, yet unstable
